tests/integration/snapshot/v3_snapshot_test.go

```diff
@@ -286,10 +286,23 @@ func restoreCluster(t *testing.T, clusterN int, dbPath string) (
 // TODO: TLS
 func newEmbedURLs(n int) (urls []url.URL) {
 	urls = make([]url.URL, n)
+	allocatedPorts := make(map[int]struct{})
 	for i := 0; i < n; i++ {
-		rand.Seed(int64(time.Now().Nanosecond()))
-		u, _ := url.Parse(fmt.Sprintf("unix://localhost:%d", rand.Intn(45000)))
+		u, _ := url.Parse(fmt.Sprintf("unix://localhost:%d", allocateUniquePort(2000, 45000, allocatedPorts)))
 		urls[i] = *u
 	}
 	return urls
 }
+
+func allocateUniquePort(minPort, maxPort int, allocatedPort map[int]struct{}) int {
```

Comment:
Wow, just wow. How many different places we have in tests that allocate shared resources (port,unix sockets, etc)?

Instead of having this generator, can we maybe try use a `LocalListenCount` https://github.com/etcd-io/etcd/blob/main/tests/framework/integration/cluster.go#L88 or adapt https://github.com/etcd-io/etcd/blob/b00bb535f09c627c4217a9f5a22d1327c4560219/tests/framework/integration/cluster.go#L525-L530 function to return port number.

Reply:
The new added `allocateUniquePort` is only used by snapshot related test, which depends on [embed.StartEtcd](https://github.com/etcd-io/etcd/blob/b00bb535f09c627c4217a9f5a22d1327c4560219/tests/integration/snapshot/v3_snapshot_test.go#L45-L71) to start etcd directly. 

We can consider how to unify all such resources allocation in a separate & dedicate PR.

Reply:
ok, however are you sure that ports generated by `newEmbedURLs` were conflicting with itself or maybe with ports generated by `newLocalListener` ?

I want to make sure we are making a proper improvement, not just continue with band-aid solutions.

Reply:
Yes, basically I am sure about it based on error message. The call stack is simple and clear.

Reply:
Another option for generator is to create a pool of ports guaranteed to be unused using: 
  a) creation of listeners with port 0 (https://cs.opensource.google/go/go/+/refs/tags/go1.18.1:src/net/http/httptest/server.go;l=68)
  b) getting the port (https://cs.opensource.google/go/go/+/refs/tags/go1.18.1:src/net/http/httptest/server.go;l=130#:~:text=129-,130,-131)
  c) closing the listener

Reply:
Yes, it's a feasible solution to unify the port allocation. Let's address it in a separate & dedicate PR.

Reply:
Updated per [discussion_r857364543](https://github.com/etcd-io/etcd/pull/13982#discussion_r857364543).  PTAL @ptabor @serathius 

With regard to unify the port allocation such as [cluster.go#L525-L530](https://github.com/etcd-io/etcd/blob/b00bb535f09c627c4217a9f5a22d1327c4560219/tests/framework/integration/cluster.go#L525-L530) , I tend to keep it as it's for now, because the existing behavior is easier to debug for test failure, because the port number is more predictable. Anyway, let's address it in a separate PR.